adapters/liftoff/liftoff.go

```diff
@@ -0,0 +1,180 @@
+package liftoff
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/prebid/openrtb/v19/openrtb2"
+	"github.com/prebid/prebid-server/adapters"
+	"github.com/prebid/prebid-server/config"
+	"github.com/prebid/prebid-server/errortypes"
+	"github.com/prebid/prebid-server/openrtb_ext"
+)
+
+const SupportedCurrency = "USD"
+
+type adapter struct {
+	Endpoint string
+}
+
+type prebid struct {
+	Debug bool `json:"debug"`
+}
+
+type prebidExt struct {
+	Prebid prebid `json:"prebid"`
+}
+
+type liftoffImpressionExt struct {
+	*adapters.ExtImpBidder
+	// Ext represents the vungle extension.
+	Ext openrtb_ext.ImpExtLiftoff `json:"vungle"`
+}
+
+/* Builder */
+
+func Builder(bidderName openrtb_ext.BidderName, config config.Adapter, server config.Server) (adapters.Bidder, error) {
+	return &adapter{Endpoint: config.Endpoint}, nil
+}
+
+/* MakeRequests */
+
+func (a *adapter) MakeRequests(request *openrtb2.BidRequest, requestInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	var requests []*adapters.RequestData
+	var errs []error
+	requestCopy := *request
+	for _, imp := range request.Imp {
+		// check if video ad
+		if imp.Video == nil {
+			errs = append(errs, fmt.Errorf("liftoff adapter handles video ads only"))
+			continue
+		}
+
+		// Check if imp comes with bid floor amount defined in a foreign currency
+		if imp.BidFloor > 0 && imp.BidFloorCur != "" && strings.ToUpper(imp.BidFloorCur) != SupportedCurrency {
+			// Convert to US dollars
+			convertedValue, err := requestInfo.ConvertCurrency(imp.BidFloor, imp.BidFloorCur, SupportedCurrency)
+			if err != nil {
+				errs = append(errs, fmt.Errorf("failed to convert currency (err)%s", err.Error()))
+				continue
+			}
+
+			// Update after conversion. All imp elements inside request.Imp are shallow copies
+			// therefore, their non-pointer values are not shared memory and are safe to modify.
+			imp.BidFloorCur = SupportedCurrency
+			imp.BidFloor = convertedValue
+		}
+
+		var impExt liftoffImpressionExt
+		if err := json.Unmarshal(imp.Ext, &impExt); err != nil {
+			errs = append(errs, fmt.Errorf("failed unmarshalling imp ext (err)%s", err.Error()))
+			continue
+		}
+
+		// get bid token & IDs
+		var bidderImpExt openrtb_ext.ImpExtLiftoff
+		if err := json.Unmarshal(impExt.Bidder, &bidderImpExt); err != nil {
+			errs = append(errs, fmt.Errorf("failed unmarshalling bidder imp ext (err)%s", err.Error()))
+			continue
+		}
+
+		if bidderImpExt.PubAppStoreID != "" && bidderImpExt.PlacementRefID != "" && bidderImpExt.BidToken != "" {
+			impExt.Ext = bidderImpExt
+			if newImpExt, err := json.Marshal(impExt); err == nil {
+				imp.Ext = newImpExt
+			} else {
+				errs = append(errs, fmt.Errorf("failed re-marshalling imp ext"))
```

Comment:
Please use errors.New instead of fmt.Errorf when there are no variables / macros in the error string.